device/esp_tinyusb/tinyusb_msc.c

```diff
@@ -0,0 +1,806 @@
+{
+    (void) lun;
+    (void) power_condition;
+
+    if (load_eject && !start) {
+        // Eject media from the storage
+        msc_storage_mount_to_app();
+    }
+    return true;
+}
+
+// Invoked when received SCSI READ10 command
+// - Address = lba * BLOCK_SIZE + offset
+// - Application fill the buffer (up to bufsize) with address contents and return number of read byte.
+int32_t tud_msc_read10_cb(uint8_t lun, uint32_t lba, uint32_t offset, void *buffer, uint32_t bufsize)
+{
+    esp_err_t err = msc_storage_read_sector(lba, offset, bufsize, buffer);
+    if (err != ESP_OK) {
+        ESP_LOGE(TAG, "msc_storage_read_sector failed: 0x%x", err);
+        return 0;
+    }
+    return bufsize;
+}
+
+/**
+ * @brief Handles deferred USB MSC write operations.
+ *
+ * This function is invoked via TinyUSB's deferred execution mechanism to perform
+ * write operations to the underlying storage. It writes data from the
+ * `storage_buffer` stored within the `s_storage_handle`.
+ *
+ * @param param Unused. Present for compatibility with deferred function signature.
+ */
+static void storage_write_func(void *param)
+{
+    assert(param); // Ensure param is not NULL
+    // Process the data in storage_buffer
+    msc_storage_obj_t *s_storage_handle = p_msc_driver->dynamic.storage;
+    esp_err_t err = msc_storage_write_sector(
+                        s_storage_handle->storage_buffer.lba,
+                        s_storage_handle->storage_buffer.offset,
+                        s_storage_handle->storage_buffer.bufsize,
+                        (const void *)s_storage_handle->storage_buffer.data_buffer
+                    );
+    if (err != ESP_OK) {
+        ESP_LOGE(TAG, "Write failed, error=0x%x", err);
+    }
+}
+
+// Invoked when received SCSI WRITE10 command
+// - Address = lba * BLOCK_SIZE + offset
+// - Application write data from buffer to address contents (up to bufsize) and return number of written byte.
+int32_t tud_msc_write10_cb(uint8_t lun, uint32_t lba, uint32_t offset, uint8_t *buffer, uint32_t bufsize)
+{
+    // There is no way to return the error from the deferred function, so we need to check everything here
+    if (bufsize > MSC_STORAGE_BUFFER_SIZE) {
+        ESP_LOGE(TAG, "Buffer size %"PRIu32" exceeds maximum allowed size %d", bufsize, MSC_STORAGE_BUFFER_SIZE);
+        goto error;
+    }
+
+    size_t addr = 0; // Address of the data to be read, relative to the beginning of the partition.
+    size_t temp = 0;
+    size_t sector_size = p_msc_driver->dynamic.storage->sector_size;
+
+    if (__builtin_umul_overflow(lba, sector_size, &temp)) {
+        ESP_LOGE(TAG, "WRITE(10): Overflow lba");
+        goto error;
+    }
+
+    if (__builtin_uadd_overflow(temp, offset, &addr)) {
+        ESP_LOGE(TAG, "WRITE(10): Overflow addr");
+        goto error;
+    }
+
+    if (addr % sector_size != 0 || bufsize % sector_size != 0) {
+        ESP_LOGE(TAG, "WRITE(10): Invalid arguments");
+        goto error;
+    }
```

Comment:
We need to verify overflow, only for SPI Flash storage.
So it is better to return the internal storage member storage_type and set it to TINYUSB_MSC_STORAGE_TYPE_SPIFLASH during the storage creation. 

Based on that, verify the overflow and report the error with status: 
```suggestion
    if (s_storage_handle->storage_type == TINYUSB_MSC_STORAGE_TYPE_SPIFLASH) {
        size_t addr = 0; // Address of the data to be read, relative to the beginning of the partition.
        size_t temp = 0;
        size_t sector_size = p_msc_driver->dynamic.storage->sector_size;

        if (__builtin_umul_overflow(lba, sector_size, &temp)) {
            ESP_LOGE(TAG, "WRITE(10): Overflow address");
            goto error;
        }

        if (__builtin_uadd_overflow(temp, offset, &addr)) {
            ESP_LOGE(TAG, "WRITE(10): Overflow offset");
            goto error;
        }

        if (addr % sector_size != 0 || bufsize % sector_size != 0) {
            ESP_LOGE(TAG, "WRITE(10): Invalid arguments");
            goto error;
        }
    }
```